apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -18,30 +17,66 @@ defmodule Aecore.Peers.Sync do
   end
 
   def init(state) do
-    Process.send_after(self(), :work, 5_000)
     {:ok, state}
   end
 
-  def handle_info(:work, state) do
-    check_peers()
-    introduce_variety()
-    refill()
-    schedule_work()
-    {:noreply, state}
+  def get_state() do
+    GenServer.call(__MODULE__, :get_state)
   end
 
-  defp schedule_work() do
-    Process.send_after(self(), :work, @check_time)
+  @spec add_block_to_state(binary(), term()) :: :ok
+  def add_block_to_state(block_hash, block) do
+    GenServer.call(__MODULE__, {:add_block_to_state, block_hash, block})
   end
 
-  defp check_peers do
-    Peers.check_peers()
+  @spec ask_peers_for_unknown_blocks(map()) :: :ok
+  def ask_peers_for_unknown_blocks(peers) do
+    GenServer.call(__MODULE__, {:ask_peers_for_unknown_blocks, peers})
+  end
+
+  @spec add_valid_peer_blocks_to_chain() :: :ok
+  def add_valid_peer_blocks_to_chain() do
+    GenServer.call(__MODULE__, :add_valid_peer_blocks_to_chain)
+  end
+
+  def handle_call(:get_state, _from, state) do
+    {:reply, state, state}
+  end
+
+  def handle_call({:add_block_to_state, block_hash, block}, _from, state) do
+    updated_state =
+      case Chain.has_block?(Base.encode16(block_hash)) do
+        true ->
+          state
+        false ->
```

Comment:
We could move block_single_validate hear so we never store not validated blocks in memory.